docs/Heterogeneous_clusters.rst

```diff
@@ -0,0 +1,82 @@
+
+For some time now new instruction sets are how the most significant performance differences in new architectures
+are realized. The most common example is the width and operations of vectorization extensions (e.g. SSE, AVX),
+and so building software that takes advantage of those is crucial for HPC.
+
+.. contents::
+    :depth: 3
+    :backlinks: none
+
+.. _heterogenous_clusters_defaults:
+
+Default behaviour of EasyBuild in heterogeneous clusters
+--------------------------------------------------------
+
+By default, EasyBuild optimizes builds for the CPU architecture of the build host, by instructing the compiler to
+generate instructions for the highest instruction set supported by the process architecture of the build host
+processor (cfr. :ref:`controlling_compiler_optimization_flags`).
+
+In an heterogenous cluster, this means that the software may not run in nodes that do not support the build host's
+instruction set (it would exit with an ``Illegal instruction`` error, in the case of software built with GNU toolchains,
+or ``Please verify that both the operating system and the processor support X, Y and Z instructions`` for software built
+with Intel toolchains), and that it will not be fully optimized when running in nodes that support higher instruction
+sets than those of the build host. The first problem can be solved by building with ``--optarch=GENERIC``, but it will
+make the second problem even worse.
+
+(With an Intel toolchain, the problem can be reduced by generating multiple code paths with the ``-ax`` compiler option
+in ``--optarch`` and by leveraging on MKL's automatic dispatch according to the execution node's instruction set, but no
+such option is available (yet) in the GNU toolchains)
+
+The solution is	then to	build multiple copies of each software,	at least for those where performance is	crucial, which is
+easily achieved simply by running EasyBuild from each type of node, the caveat being where exactly to install copies for
+different architectures in a way that they can be loaded, with their dependencies, and used effectively across the cluster
+by all users.
+
+.. _heterogeneous_clusters_visibility:                                                                                                                                                                                                         
+Visibility of achitectures in heterogenous clusters
+---------------------------------------------------
+                                                           
+One way of distinguishing between the many alternatives for using EasyBuild in an heterogeneous cluster concerns whether
+each host only sees the software compiled for its own architecture (plus any software eventually compiled for ``GENERIC``)
+or if it sees everything.
+
+By mounting architecture dependent targets on the same mountpoint in every host, the configuration is then very similar to
```

Comment:
Would it only be changing the mointpoint? We thought about directories like /common-path/easybuild/<architecture>/ where architecture is the instruction set name of gcc (nehalem, broadwell, skylake,...) to start with. On the nodes one would change the profile.d to prepend $MODULEPATH with the according modules path, hence module load would pick up the more optimized modules first and fall back to the other architectures, i.e. instruction sets for other modules. For example, MODULEPATH=/<....>/skylake/modules/all:/<....>/nehalem/modules/all.

Reply:
@henkela this clearly needs to be rewritten, my goal was precisely to describe those two approaches (a node only seeing it's own arch, via mountpoints, or seeing everything but using only it's own arch, via variables).

Using variables is of course much more flexible, but in many simple cases using mountpoints may be more robust and foolproof (from the point of view the nodes, except for the custom mountpoint, it's the same as an homogeneous cluster)